examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,839 @@
+            /*.mem_size   =*/ n_bytes + ggml_tensor_overhead(),
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ false,
+        };
+        struct ggml_context * ctx_data = ggml_init(params);
+
+        printf("inp_pos [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(inp_pos, 0, DEBUG_POS, 0, 0));
+        printf("inp_neg [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(inp_neg, 0, DEBUG_POS, 0, 0));
+
+        // TODO is this the best way to get dimension? i don't know which way n_embd/n_tokens go
+        // for that matter can we get rid of n_embd/n_tokens fields in favor of ne[0]/ne[1]?
+        struct ggml_tensor * dest = ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, inp_pos->ne[0], inp_pos->ne[1]);
+    
+        for (size_t i = 0; i < cb_data.n_embd; i++) {
+            for (size_t j = 0; j < cb_data.n_tokens; j++) {
+                ggml_set_f32_nd(dest, i, j, 0, 0, ggml_get_f32_nd(inp_pos, i, j, 0, 0) - ggml_get_f32_nd(inp_neg, i, j, 0, 0));
+            }
+        }
+
+        printf("dest [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(dest, 0, DEBUG_POS, 0, 0));
+
+        // TODO can we make this faster? like check during the above operation rather than on a second pass?
+
+        // strip zero rows
+        std::vector<int> nonzero_rows;
+        for (int i = 0; i < cb_data.n_tokens; ++i) {
+            if (!is_row_all_zeros(dest, i, cb_data.n_embd)) {
+                nonzero_rows.push_back(i);
+            }
+        }
+
+        /* debug
+        if(cb_data.n_tokens != nonzero_rows.size()) {
+            std::cout << "original n_tokens: " << cb_data.n_tokens << std::endl;
+            std::cout << "zero rows in layer " << il << ": " << cb_data.n_tokens - nonzero_rows.size() << std::endl;
+        } */
+
+        struct ggml_init_params params2 = {
+            /*.mem_size   =*/ inp_pos->ne[0] * nonzero_rows.size() * sizeof(float) + ggml_tensor_overhead(),
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ false,
+        };
+        struct ggml_context * ctx_data2 = ggml_init(params);
+
+        struct ggml_tensor * diff = ggml_new_tensor_2d(ctx_data2, GGML_TYPE_F32, inp_pos->ne[0], nonzero_rows.size());
+
+        //size_t offset = 0;
+        for (size_t i = 0; i < nonzero_rows.size(); ++i) {
+            // probably eschew this in favor of the iterative method?
+            //float * origin = (float *)(dest->data) + nonzero_rows[i] * cb_data.n_embd;
+            //memcpy((float *)(diff->data) + offset, origin, cb_data.n_embd * sizeof(float));
+            //offset += cb_data.n_embd;
+
+            for (size_t j = 0; j < cb_data.n_embd; j++) {
+                ggml_set_f32_nd(diff, j, i, 0, 0, ggml_get_f32_nd(dest, j, nonzero_rows[i], 0, 0));
+            }
+        }
+
+        // FIXME ggml_nbytes(diff) is 0
+
+        cb_data.v_diffs_wrapped[il].push_back(diff);
+        ggml_free(ctx_data);
+        ggml_free(ctx_data2);
+    }
+}
+
+static void concatenate_diffs(callback_data & cb_data) {
```

Comment:
Because the values are (or will be in my intended fix) at adjacent memory locations, is there any way to more quickly concatenate `ggml_tensor`s? I'd like to do it inplace without having to deal with another `ggml_context`.

Reply:
One idea is that we could re-use the v_diff tensor, just need to reallocate memory. Something like this:

```cpp
void * v_diff_tensor_data = v_diff_tensor->data; // save current value to temporary variable
v_diff_tensor->data = malloc(...); // alloc a big memory for the final v_diff tensor
memcpy(...); // copy & concat the v_diff tensors
```

Or another idea (better one) is to write the diff vector to the correct offset right inside `calc_diff`. But for now it's not very clear in my head how we tell `calc_diff` what is the offset to write data into.